lib/live_view_native/extensions/modifiers.ex

```diff
@@ -6,41 +6,271 @@ defmodule LiveViewNative.Extensions.Modifiers do
+  modifier builder stack through its implementation of `LiveViewNativePlatform.ModifiersStack`.
+  The modifier builder is then returned with the modifier added to the stack.
+
+  ```heex
+  <Text modifiers={some_other_modifier() |> bold(%{is_active: true})}>And even this one as well.</Text>
+  ```
+
+  This example is the same as the previous one except that the second value is an option
+  list. The same steps are applied as the second example, with the same return value being
+  added to the passed modifier builder's stack.
+
+  The "modifier builder" that gets returned in all of these examples is a custom struct
+  defined by the platform library that provides the modifier function being called. In all
+  cases, it is a module that implements the following protocols:
+
+  - `LiveViewNativePlatform.ModifiersStack`
+  - `Jason.Encoder`
+  - `Phoenix.HTML.Safe`
+
+  These implementations are used for appending modifiers, encoding them as JSON and rendering
+  them as HTML attributes respectively. For platforms that don't provide a modifier builder
+  struct, a `%LiveViewNativePlatform.GenericModifiers{}` is used as a fallback.
   """
+  alias LiveViewNativePlatform.ModifiersStack
+
   defmacro __using__(opts \\ []) do
     quote bind_quoted: [
             custom_modifiers: opts[:custom_modifiers],
+            modifiers_struct: opts[:modifiers_struct],
             platform_modifiers: opts[:platform_modifiers],
             platform_module: opts[:platform_module]
           ] do
       all_modifiers = Keyword.merge(platform_modifiers, custom_modifiers)
 
       if is_nil(platform_module) do
         for {modifier_key, modifier_module} <- all_modifiers do
-          def unquote(:"#{modifier_key}")(ctx, params \\ %{}) do
-            modifiers = ctx.modifiers
+          # When a function with the name `params` is defined on a modifier module, a function
+          # with the same arity and the name of the modifier is generated. This function takes
+          # the same arguments as the `params` function, calling that function before applying
+          # a modifier to the stack.
+
+          def unquote(:"#{modifier_key}")({:_apply_mod, {mod_args, mod_builder, opts}}) do
+            mod_builder = mod_builder || struct(unquote(modifiers_struct), %{})
+            raw_params = Keyword.get(opts, :raw_params)
+            params = if raw_params, do: raw_params, else: apply(unquote(modifier_module), :params, mod_args)
             modifier = apply(unquote(modifier_module), :modifier, [params])
 
-            if modifiers do
-              Map.put(
-                ctx,
-                :modifiers,
-                LiveViewNativePlatform.ModifiersStack.append(modifiers, modifier)
-              )
-            else
-              ctx
+            ModifiersStack.append(mod_builder, modifier)
+          end
+
+          if Keyword.has_key?(modifier_module.__info__(:functions), :params) do
```

Comment:
Any chance this block could be auto-generated? Seems like something that could be accidentally broken.

Reply:
@carson-katri Hmm, I considered that. The tricky part is dynamically generating the function head with the correct arity for an arbitrary `n` which I'm not quite sure how to do with Elixir metaprogramming. I do agree that this is prone to human error though, so I'll look into this or an alternative implementation that avoids the repetition (if you have any suggestions, let me know).